java/code/src/com/suse/manager/gatherer/HostJson.java

```diff
@@ -91,23 +92,23 @@ public String getHostIdentifier() {
      * @return the number of sockets
      */
     public Integer getTotalCpuSockets() {
-        return totalCpuSockets;
+        return Optional.ofNullable(totalCpuSockets).orElse(1);
```

Comment:
Two points here, one nitpicking and one deeper question.

Starting with the nitpicking :slightly_smiling_face: but for these type of uses cases I'd rather use a simple `== null` check or `Objects.requireNonNullElse()` if we want to have a more fluent syntax. Using an `Optional` in this way means we are creating an object in the heap just to throw it away immediately. For this specific case it probably doesn't matter, but it does introduces a significant memory overhead if we use this technique for parsing large JSON files.

The other point is related to the value we are using here as default. This `totalCpuSockets` field is stored as `nrSockets` in the `CPU` object [by the VirtualHostManagerProcessor](https://github.com/uyuni-project/uyuni/blob/c3aa9878e05f6f0e4e31e4086a8926f7cc85aaea/java/code/src/com/redhat/rhn/taskomatic/task/gatherer/VirtualHostManagerProcessor.java#L281). Later, this value is passed to subscription matcher as number of cpus [by the MatcherJsonIO class](https://github.com/uyuni-project/uyuni/blob/c3aa9878e05f6f0e4e31e4086a8926f7cc85aaea/java/code/src/com/suse/manager/matcher/MatcherJsonIO.java#L163). There is a Drools rule where [null values for the cpu count are translated to 16](https://github.com/uyuni-project/subscription-matcher/blob/master/src/main/resources/com/suse/matcher/rules/drools/InputValidation.drl#L116-L127).  I'm not sure if before this PR the `null` value actually reached subscription-matcher or if the code would fail earlier, but we might change the results of subscription-matcher in these corner cases.

So the question is, is 1 really a good default value for this value?

Reply:
good catch. I have changed this and use for socket value "0".
In case of parsing errors, I do not set any CPU value to keep the null for subscription matcher.
The problem is, that we get now total values from gatherer and we need to divide by sockets to get single numbers.
`null.longValue()` generate a NPE while also divide by zero also not work. So we better set nothing if we get bad data. For cores and threads I assume "1" which should not generate problems.